examples/rust/bzsstack/src/main.rs

```diff
@@ -0,0 +1,201 @@
+        }
+    } else {
+        SymbolFileCfg::Process { pid: Some(pid) }
+    };
+
+    let syms = symbolizer.symbolize(&[cfg], stack);
+    for i in 0..stack.len() {
+        if syms.len() == stack.len() {
+            if let Some(sym) = &syms[i] {
+                if !sym.path.is_empty() {
+                    println!(
+                        "  {} [<{:016x}>] {}+0x{:x} {}:{}",
+                        i,
+                        stack[i],
+                        sym.symbol,
+                        stack[i] - sym.start_address,
+                        sym.path,
+                        sym.line_no
+                    );
+                } else {
+                    println!(
+                        "  {} [<{:016x}>] {}+0x{}",
+                        i,
+                        stack[i],
+                        sym.symbol,
+                        stack[i] - sym.start_address
+                    );
+                }
+                continue;
+            }
+        }
+        println!("  {} [<{:016x}]", i, stack[i]);
+    }
+}
+
+unsafe extern "C" fn event_handler(
+    ctx: *mut ::std::os::raw::c_void,
+    data: *mut ::std::os::raw::c_void,
+    size: libbpf_sys::size_t,
+) -> ::std::os::raw::c_int {
+    let symbolizer = &*(ctx as *const BlazeSymbolizer);
+    let event = &*(data as *const stacktrace_event);
+
+    if size as usize != mem::size_of::<stacktrace_event>() {
+        eprintln!(
+            "Invalid size {} != {}",
+            size,
+            mem::size_of::<stacktrace_event>()
+        );
+        return 1;
+    }
+
+    if event.kernel_stack_size == 0 && event.user_stack_size == 0 {
+        return 1;
+    }
+
+    println!("PID {}, CPU {}", event.pid, event.cpu_id);
+
+    if event.kernel_stack_size > 0 {
+        println!("Kernel:");
+        show_stack_trace(
+            &event.kernel_stack[0..(event.kernel_stack_size as usize)],
+            symbolizer,
+            0,
+        );
+    } else {
+        println!("No Kernel Stack");
+    }
+
+    if event.user_stack_size > 0 {
+        println!("Userspace:");
+        show_stack_trace(
+            &event.user_stack[0..(event.user_stack_size as usize)],
+            symbolizer,
+            event.pid,
+        );
+    } else {
+        println!("No Userspace Stack");
+    }
+
+    println!();
+    0
+}
+
+#[derive(Parser, Debug)]
+struct Args {
+    /// Sampling frequency
+    #[clap(short, default_value_t = 1)]
+    freq: u64,
+}
+
+fn main() -> Result<(), Error> {
+    let args = Args::parse();
+    let freq = if args.freq < 1 { 1 } else { args.freq };
+
+    let mut symbolizer = BlazeSymbolizer::new()?;
+
+    let skel_builder = ProfileSkelBuilder::default();
+    let open_skel = skel_builder.open().unwrap();
+    let mut skel = open_skel.load().unwrap();
+
+    let pefds = init_perf_monitor(freq);
+    let _links = attach_perf_event(&pefds, skel.progs_mut().profile());
+
+    let ring_buf = unsafe {
+        libbpf_sys::ring_buffer__new(
```

Comment:
libbpf-rs has Rust bindings for ringbuf, you shouldn't need to have so much unsafe code.

@insearchoflosttime, can you please help reviewing Rust parts and suggesting best libbpf-rs usage? Thanks!

Reply:
I just found a problem with RingBuffer API.  RingBuffer::poll() accepts only Duration.  While ring_buffer__poll() waits indefinitely if no event happens, for passing -1 as the value of timeout, Duration can not be negative.

Reply:
bootstrap example specifies 100ms as a timeout, it doesn't hurt. But @insearchoflosttime, probably something to consider supporting to specify "no timeout" ?

Reply:
Yes, we should support that for RingBuffer::poll(). I will make sure that gets fixed in libbpf-rs

Reply:
Use Duration::MAX instead.